### PR TITLE
Register block-builder target in mimir

### PR DESCRIPTION
**Pull request by narqo**
#### What this PR does

~~_This one seats atop https://github.com/grafana/mimir/pull/9199 for now_~~

This is part of https://github.com/grafana/mimir/pull/8635; refer to it for more details.

Here we register the `block-builder` target in mimir. The block-builder is still in an earlier experiment, thus don't expose any documentation or changelog for now.

Also, the PR addresses the discussion we had with @dimitarvdimitrov in https://github.com/grafana/mimir/pull/9199#discussion_r1773600398 about a potential edge-case for a partition that has offset gaps. Here we refactor the internals of the consumer loop replacing the use of remaining lag with the reference to the partition end, at the time of the cycle.

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by narqo** [comment]
Close in favour of #9315

File: pkg/blockbuilder/blockbuilder.go [hunk 1]
```diff
@@ -289,7 +288,7 @@ func partitionStateFromLag(logger log.Logger, lag kadm.GroupMemberLag, fallbackM
 
 // consumePartition consumes records from the given partition until the cycleEnd timestamp.
 // If the partition is lagging behind, it takes care of consuming it in sections.
-func (b *BlockBuilder) consumePartition(ctx context.Context, partition int32, state partitionState, cycleEnd time.Time) (err error) {
+func (b *BlockBuilder) consumePartition(ctx context.Context, partition int32, state partitionState, cycleEnd time.Time, cycleEndOffset int64) (err error) {
```
Comment (pracucci) [inline pkg/blockbuilder/blockbuilder.go:291]: [nit] No we have `cycleEndOffset`. WDYT about renaming `cycleEnd` to something like `cycleEndTime` to disambiguate? If you accept this comment, then I would also rename `sectionEnd` accordingly.

```suggestion
func (b *BlockBuilder) consumePartition(ctx context.Context, partition int32, state partitionState, cycleEndTime time.Time, cycleEndOffset int64) (err error) {
```

**Review by pracucci — APPROVED** [review]
I'm very glad you got rid of `remaining` offsets. LGTM.

> Also, the PR addresses the discussion we had with @dimitarvdimitrov in https://github.com/grafana/mimir/pull/9199#discussion_r1773600398 about a potential edge-case for a partition that has offset gaps.

Can we unit test it? Or is it too complicated because we can't create gaps in the mocked Kafka cluster?

File: pkg/mimir/mimir.go [hunk 2]
```diff
@@ -123,6 +124,7 @@ type Config struct {
 	Worker           querier_worker.Config           `yaml:"frontend_worker"`
 	Frontend         frontend.CombinedFrontendConfig `yaml:"frontend"`
 	IngestStorage    ingest.Config                   `yaml:"ingest_storage"`
+	BlockBuilder     blockbuilder.Config             `yaml:"block_builder" doc:"hidden"`
```
Comment (dimitarvdimitrov) [inline pkg/mimir/mimir.go:127]: isn't this part of ingest storage? i mean we use the same kafka config and it's rightly coupled with kafka just like ingest storage

Comment (dimitarvdimitrov) [inline pkg/mimir/mimir.go:127]: why hidden? Ingest storage is just experimental

Comment (pracucci) [reply pkg/mimir/mimir.go:127]: I think it's fair having a root leve config. Ingest storage is the config shared between component. This config is block-builder specific.

Comment (narqo) [reply pkg/mimir/mimir.go:127]: I don't feel strong about it, but I am not comfortable exposing anything about block-builder to the user-visible helps. Block-builder is even more experimental than ingest-storage.

Comment (narqo) [reply pkg/mimir/mimir.go:127]: +1 to what Marco says. For me, ingest-storage is a package, that services like ingester, query-frontend, ruler, and now block-builder use. I.e. block-builder is not a feature of ingest-storage, it's the opposite 🤷

**Comment by narqo** [comment]
> Can we unit test it? Or is it too complicated because we can't create gaps in the mocked Kafka cluster?

For now, I cannot think of a way to replicate that in a controlled manner, so will need some exploration. I will add it to the rest of the "more tests" PRs later.
